### PR TITLE
Allow branch names longer than 7 characters

### PR DESCRIPTION
**Pull request by samuelson**
Resolves https://github.com/gogits/gogs/issues/3743

This fixed the issue for me, but I don't know if it will have some other unintended consequences.

### CODE REVIEW

**Comment by samuelson** [comment]
@Unknwon please let me know what I need to change for this to be merged.

**Comment by binford2k** [comment]
@Unknwon ping, this would be really nice to get corrected.

File: cmd/web.go [hunk 1]
```diff
@@ -585,12 +585,12 @@ func runWeb(ctx *cli.Context) error {
...
 			m.Get("/commits/*", repo.RefCommits)
-			m.Get("/commit/:sha([a-z0-9]{7,40})$", repo.Diff)
+			m.Get("/commit/:sha([a-z0-9]{40})$", repo.Diff)
 			m.Get("/forks", repo.Forks)
 		}, context.RepoRef())
-		m.Get("/commit/:sha([a-z0-9]{7,40})\\.:ext(patch|diff)", repo.RawDiff)
+		m.Get("/commit/:sha([a-z0-9]{40})\\.:ext(patch|diff)", repo.RawDiff)
 
-		m.Get("/compare/:before([a-z0-9]{7,40})\\.\\.\\.:after([a-z0-9]{7,40})", repo.CompareDiff)
+		m.Get("/compare/:before([a-z0-9]{40})\\.\\.\\.:after([a-z0-9]{40})", repo.CompareDiff)
```
Comment (0xbaadf00d) [inline cmd/web.go]: I agree. It's the root of your problem.

File: cmd/web.go [hunk 2]
```diff
@@ -585,12 +585,12 @@ func runWeb(ctx *cli.Context) error {
 			m.Get("/src/*", repo.Home)
 			m.Get("/raw/*", repo.SingleDownload)
 			m.Get("/commits/*", repo.RefCommits)
-			m.Get("/commit/:sha([a-z0-9]{7,40})$", repo.Diff)
+			m.Get("/commit/:sha([a-z0-9]{40})$", repo.Diff)
 			m.Get("/forks", repo.Forks)
 		}, context.RepoRef())
-		m.Get("/commit/:sha([a-z0-9]{7,40})\\.:ext(patch|diff)", repo.RawDiff)
+		m.Get("/commit/:sha([a-z0-9]{40})\\.:ext(patch|diff)", repo.RawDiff)
```
Comment (0xbaadf00d) [inline cmd/web.go]: This modification don't fix your issue. But the regexp is not correct : SHA1 only accept _a b c d e f 0 1 2 3 4 5 6 7 8 9_ characters. Could you replace `:sha([a-z0-9]{7,40})` with `:sha([a-f0-9]{7,40})` please.

File: cmd/web.go [hunk 3]
```diff
@@ -585,12 +585,12 @@ func runWeb(ctx *cli.Context) error {
 			m.Get("/src/*", repo.Home)
 			m.Get("/raw/*", repo.SingleDownload)
 			m.Get("/commits/*", repo.RefCommits)
-			m.Get("/commit/:sha([a-z0-9]{7,40})$", repo.Diff)
+			m.Get("/commit/:sha([a-z0-9]{40})$", repo.Diff)
```
Comment (0xbaadf00d) [inline cmd/web.go]: This modification don't fix your issue. In fact this route is not used when you call URL like `http://domain.local/zero-x-baadf00d/test/compare/production...production`

**Comment by 0xbaadf00d** [comment]
Ping @samuelson

**Comment by samuelson** [comment]
@0xbaadf00d sorry for the delay, it turns out I had notifications disabled.  I updated with the changes you suggested.

**Comment by strk** [comment]
@samuelson please consider sending this PR to https://github.com/go-gitea/gitea

**Comment by samuelson** [comment]
@strk looks like it's already fixed: https://github.com/go-gitea/gitea/blob/master/cmd/web.go#L592

**Comment by samuelson** [comment]
This is the same, go ahead an merge that one: https://github.com/gogits/gogs/pull/3851

**Pull request closed without merging**